### PR TITLE
bpo-34638: Store a weak reference to stream reader to break strong references loop

### PR DESCRIPTION
**Pull request by asvetlov**
Breaking the strong reference loop between reader and protocol 
allows to detect and close the socket if the stream is deleted (garbage collected) without `close()` call.

https://bugs.python.org/issue34638

### CODE REVIEW

File: Lib/asyncio/streams.py [hunk 1]
```diff
@@ -186,46 +189,110 @@ class StreamReaderProtocol(FlowControlMixin, protocols.Protocol):
...
+        # connection_lost() is not called yet
+        assert self._stream_reader_wr is not None
+
+        transport = self._transport
+        if transport is not None:
+            # connection_made was called
+            context = {
+                'message': ("Close transport. "
+                            "A stream was destroyed without "
+                            "stream.close() call")
```
Comment (1st1) [inline Lib/asyncio/streams.py]: I'd rephrase as:

```
An open stream is being garbage collected; 
call "stream.close()" explicitly.
```

File: Lib/asyncio/streams.py [hunk 2]
```diff
@@ -186,46 +189,110 @@ class StreamReaderProtocol(FlowControlMixin, protocols.Protocol):
...
+            return None
+        return self._stream_reader_wr()
+
     def connection_made(self, transport):
-        self._stream_reader.set_transport(transport)
+        if self._reject_transport:
+            context = {
+                'message': ("Close transport. "
+                            "a stream was destroyed "
+                            "before connection establishment")
```
Comment (1st1) [inline Lib/asyncio/streams.py]:
```
An open stream was garbage collected prior to 
establishing network connection; call "stream.close()" explicitly.
```

File: Lib/asyncio/streams.py [hunk 3]
```diff
@@ -282,6 +349,9 @@ def can_write_eof(self):
         return self._transport.can_write_eof()
 
     def close(self):
+        # a reader could be garbage collected / destroyed
+        # after connection closing
```
Comment (1st1) [inline Lib/asyncio/streams.py]:
```
# The reader can be safely garbage collected after 
# closing the connection.
```

File: Lib/asyncio/streams.py [hunk 4]
```diff
@@ -186,46 +189,107 @@ class StreamReaderProtocol(FlowControlMixin, protocols.Protocol):
...
+            self._stream_reader_wr = weakref.ref(stream_reader,
+                                                 self._on_reader_gc)
+            self._source_traceback = stream_reader._source_traceback
+        else:
+            self._stream_reader_wr = None
+        if client_connected_cb is not None:
+            # server socket
+            # we need to keep a strong reference to the reader
+            # until connection is made
+            self._strong_reader = stream_reader
```
Comment (1st1) [inline Lib/asyncio/streams.py]:
```
# This is a stream created by the `create_server()` function.
# Keep a strong reference to the reader until a connection 
# is established.
```

**Comment by asvetlov** [comment]
Notes are fixed

File: Lib/asyncio/subprocess.py [hunk 5]
```diff
@@ -36,6 +36,11 @@ def __repr__(self):
             info.append(f'stderr={self.stderr!r}')
         return '<{}>'.format(' '.join(info))
 
+    def _untrack_reader(self):
+        # a dummy placeholder subprocess protocol doesn't need it but
+        # the method is required for regular stream implementation.
+        pass
```
Comment (1st1) [inline Lib/asyncio/subprocess.py]:
```
`StreamWriter.close()` expects the protocol to have this method defined.
```

**Review by 1st1 — APPROVED** [review]
